### PR TITLE
Refactor test configuration

### LINKED ISSUE #155: Reconfigure test secrets so that Dependabot tests also pass
Currently many Dependabot PRs are failing the backend tests because they don't have access to the Action secrets.

image

But we cannot expose these secrets since, well, they are secrets for a reason.

I will have a look and see whether the emulator can run without certain secrets (that we must not expose) and reconfigure GitHub Actions.

Don't ask me why two PRs still managed to pass here &ndash; I don't know either.

### PR DESCRIPTION
**Pull request by joulev**
Fix #155.

Unrelated to the purpose of this PR, but I also converted a bunch of <code>it(\`abc\`)</code> to `it("abc")` in the server tests.

All actions secrets have been deleted.

`CODEOWNERS` will be updated after this is merged.

### CODE REVIEW

**Comment by joulev** [comment]
As a follow-up, please rename your local `.env` to `.env.local` to most closely mimic how the workflows are run. Reason: `.env.local` isn't exposed to tests, but `.env` is, and we want to make sure certain environment variables aren't needed in the tests.

**Comment by joulev** [comment]
Hi, review this.

**Comment by VietAnh1010** [comment]
With the current changes, `yarn test:server:ci` does not work. To replicate:

- Open a terminal session and run `firebase emulators:start --project <whatever>`
- Open another terminal session and run `yarn test:server:ci`. All tests will fail.

The reason behind this is because `applicationDefault()` cannot find the projectId, as the emulator is running in a different process.

I will fix this before we merge the branch to `main`.

File: server/firebase/firebaseAdmin.ts [hunk 1]
```diff
@@ -13,8 +13,12 @@ const serviceAccount = {
...
     ? admin.app()
     : initializeApp({
-          credential: cert(serviceAccount),
-          storageBucket: process.env.FIREBASE_STORAGE_BUCKET,
+          credential:
+              process.env.NODE_ENV === "test"
+                  ? admin.credential.applicationDefault()
+                  : cert(serviceAccount),
+          storageBucket:
+              process.env.NODE_ENV === "test" ? "test_bucket" : process.env.FIREBASE_STORAGE_BUCKET,
```
Comment (VietAnh1010) [inline server/firebase/firebaseAdmin.ts:21]: `storageBucket` for a project with `projectId` is `<projectId>.appspot.com`.

Comment (joulev) [reply server/firebase/firebaseAdmin.ts:21]: Yes I know. However it isn't necessary for the test environment.

File: package.json [hunk 2]
```diff
@@ -12,8 +12,8 @@
     "test:client:ci": "jest --ci --coverage --config=jest.client.config.js",
     "test:server": "jest --watch --coverage --config=jest.server.config.js",
     "test:server:ci": "jest --ci --coverage --config=jest.server.config.js",
-    "emulate:test:server": "firebase emulators:exec 'yarn test:server'",
-    "emulate:test:server:ci": "firebase emulators:exec 'yarn test:server:ci'"
+    "emulate:test:server": "firebase emulators:exec 'yarn test:server' --project test-proj",
```
Comment (VietAnh1010) [inline package.json:15]: I think we should use a demo project here, by prefixing the project id with `demo-`? For example: `demo-proj`.

**Review by VietAnh1010 — COMMENTED** [review]
Apart from a few configurations, other changes are not very significant.

I will resolve the configurations myself.

**Comment by joulev** [comment]
> Apart from a few configurations, other changes are not very significant.
> 
> I will resolve the configurations myself.

No, wait a minute.

File: package.json (hunk 2 above)
Comment (joulev) [reply package.json:15]: What is the difference?

**Comment by joulev** [comment]
Tell me your development workflow before this PR (what commands are run, how are they run, etc.) and I will update the Yarn scripts accordingly.

**Comment by VietAnh1010** [comment]
> This is not how development is supposed to work. You are expected to add all development workflow as Yarn scripts, and run from there. Also you shouldn't have to run two commands side-by-side. (what???)

No no no, development workflow will not be changed.

I want `yarn test:server:ci` to work because starting the emulators takes some time and hook up quite some resource on my local machine. It is not very efficient to keep running `yarn emulate:test:server:ci` repeatedly. Instead, I just need to boost the emulator once before running `yarn test:server:ci` multiple times on my machine.

With development workflow, because we just need to run the tests once, use `yarn emulate:test:server:ci` is enough.

**Comment by VietAnh1010** [comment]
Side note: it seems that the tests are run even on `push` to other branches. Is this intended?

**Comment by joulev** [comment]
> It is not very efficient to keep running yarn emulate:test:server:ci repeatedly.

Even I, with a relatively fast computer now, don't run `yarn test:client:ci`, because it is literally for CI environments. You would want to run `yarn emulate:test:server` instead, and for that the set-up step is (of course) only executed once.

Please do _not_ use `yarn (emulate:)test:server:ci` to do things that `yarn (emulate:)test:server` should be doing.

**Comment by joulev** [comment]
> Side note: it seems that the tests are run even on push to other branches. Is this intended?

Yes. I had test failures going undetected on other branches before. Now all tests should run on all branches in this repo on every commit, and that is intended.

**Comment by joulev** [comment]
Okay since this doesn't relate to frontend you can do however you want. One piece of advice though is that `:ci` variants should only be run on the CI pipelines.

**Comment by VietAnh1010** [comment]
Noted.

This branch can be merged now.

**Comment by joulev** [comment]
Okay thanks!

**Pull request merged**